scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -741,57 +766,68 @@ object Execution {
         tail,
         ExecutionOptimizationRules.MapWrite.ComposeMap(result, mapFn))
 
-    def unwrapListEither[A, B, C](it: List[(A, Either[B, C])]): (List[(A, B)], List[(A, C)]) = it match {
-      case (a, Left(b)) :: tail =>
+    private def unwrapListEither[A, B, C, D](it: List[(A, (Either[B, C], D))]): (List[(A, B, D)], List[(A, C, D)]) = it match {
+      case (a, (Left(b), d)) :: tail =>
         val (l, r) = unwrapListEither(tail)
-        ((a, b) :: l, r)
-      case (a, Right(c)) :: tail =>
+        ((a, b, d) :: l, r)
+      case (a, (Right(c), d)) :: tail =>
         val (l, r) = unwrapListEither(tail)
-        (l, (a, c) :: r)
+        (l, (a, c, d) :: r)
       case Nil => (Nil, Nil)
     }
 
     // We look up to see if any of our ToWrite elements have already been ran
     // if so we remove them from the cache.
     // Anything not already ran we run as part of a single flow def, using their combined counters for the others
     protected def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) = {
-      Trampoline(cache.getOrElseInsert(conf, this, {
-        val cacheLookup: List[(ToWrite[_], Either[Promise[Map[Long, ExecutionCounters]], Future[Map[Long, ExecutionCounters]]])] =
+      lazy val uncachedFutureCancel = {
+        val cacheLookup: List[(ToWrite[_], (Either[Promise[Map[Long, ExecutionCounters]], Future[Map[Long, ExecutionCounters]]], CancellationHandler))] =
           (head :: tail).map{ tw => (tw, cache.getOrLock(conf, tw)) }
         val (weDoOperation, someoneElseDoesOperation) = unwrapListEither(cacheLookup)
 
         val otherResult = failFastSequence(someoneElseDoesOperation.map(_._2))
+        val otherCancellationHandler = someoneElseDoesOperation.map(_._3).foldLeft(CancellationHandler.empty) { case (acc, cancel) => acc.compose(cancel) }
+
         otherResult.value match {
-          case Some(Failure(e)) => Future.failed(e)
+          case Some(Failure(e)) => (Future.failed(e), CancellationHandler.empty)
```

Comment:
should we call the `otherCancellationHandler` here? Can you add a comment in any case.